pkg/vmprovider/providers/vsphere/virtualmachine/backup.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2023 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package virtualmachine
+
+import (
+	goctx "context"
+
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/types"
+	corev1 "k8s.io/api/core/v1"
+	ctrlruntime "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+
+	vmopv1 "github.com/vmware-tanzu/vm-operator/api/v1alpha1"
+	"github.com/vmware-tanzu/vm-operator/pkg/context"
+	"github.com/vmware-tanzu/vm-operator/pkg/util"
+	"github.com/vmware-tanzu/vm-operator/pkg/vmprovider/providers/vsphere/constants"
+)
+
+// BackupVirtualMachine backs up the VM's kube data and bootstrap data into the VM's ExtraConfig.
+func BackupVirtualMachine(
+	vmCtx context.VirtualMachineContext,
+	vcVM *object.VirtualMachine,
+	k8sClient ctrlruntime.Client) error {
+	vmKubeData, err := getEncodedVMKubeData(vmCtx.VM)
+	if err != nil {
+		vmCtx.Logger.Error(err, "Failed to get VM kube data for backup")
+		return err
+	}
+
+	vmBootstrapData, err := getEncodedVMBootstrapData(vmCtx.Context, vmCtx.VM, k8sClient)
+	if err != nil {
+		vmCtx.Logger.Error(err, "Failed to get VM bootstrap data for backup")
+		return err
+	}
+	if vmBootstrapData == "" {
+		vmCtx.Logger.V(4).Info("No bootstrap data is set to the VM for backup")
+	}
+
+	_, err = vcVM.Reconfigure(vmCtx, types.VirtualMachineConfigSpec{
```

Comment:
As a followup, we can try to optimize it so we only trigger the Reconfigure if the value of the key doesn't exist, or has been updated.  IIUC, your current implementation will spew the VC UI with a Reconfigure op during each reconcile. File a story/issue if handling in a followup.

Reply:
Good point! I kind of thought about this but to confirm if there were any changes, it would require another VC call to retrieve the existing ExtraConfig, unless we store that info in memory. I will file a story internally to discuss this along with the Event v. Condition case below.

Reply:
> We now have the ability to set the ExtraConfig at VC DeployOVF time. Similarly too, many customizations result in a reconfigure anyways so we can lump this in there as well to get something closer to actual snapshot.

+1. Agreed. We can lump these into the ConfigSpec during the DeployOVF call and also only do reconfigure if the key doesn't exist as an ExtraConfig in the VM during update.

Reply:
Thanks all for the feedback on this. I have filed an internal ticket and will address it and some other improvements in a follow-up commit soon. The current implementation is guarded by a disabled FSS, so it should not affect the current VMOP functionality.